### ISSUE TITLE
Incorrect inference for instance attributes typed inside `__init__`

### ISSUE THREAD

**Issue by pkch** [python/mypy#3143]
```
class X:
    def __init__(self) -> None:
        self.link: X = X()

class Y(X):
    # link: Y
    def __init__(self) -> None:
        self.link: Y = Y()

def f(x: X) -> None:
    if isinstance(x, Y):
        reveal_type(x.link)
```

Here, the revealed type should be `Y`, but mypy says it's `X`. 

There is an easy workaround, just uncomment the line that says `link: Y`, and the inference will be correct. But since we do allow instance attribute annotations inside `__init__`, I guess it would be nice to fix.

**Comment by ilevkivskyi** [comment]
It looks like currently mypy recognizes an explicit type on ``self`` assignment only when it creates a _new_ variable. I agree this behaviour is misleading.

**Comment by gvanrossum** [comment]
Agreed. ISTM mypy should give an error whenever there is a type annotation that it isn't processing. (IIRC PEP 526 syntactically allows type annotations on places that we wouldn't expect a type checker to process -- but those should be rejected by the checker, not silently ignored.)

**Comment by AlexWaygood** [comment]
This seems to be fixed now; can't reproduce in 0.940: https://mypy-play.net/?mypy=latest&python=3.10&gist=f16b383b009d196f7165f157674aa195
